library/src/main/java/com/datatorrent/lib/join/AbstractJoinOperator.java

```diff
@@ -0,0 +1,341 @@
+  // Strategy of Join operation, by default the option is inner join
+  protected JoinStrategy strategy = JoinStrategy.INNER_JOIN;
+  // This represents whether the processing tuple is from left port or not
+  protected Boolean isLeft;
+  @InputPortFieldAnnotation(optional = true)
+  public transient DefaultInputPort<T> input1 = new DefaultInputPort<T>()
+  {
+    @Override
+    public void process(T tuple)
+    {
+      isLeft = true;
+      processTuple(tuple);
+    }
+  };
+  @InputPortFieldAnnotation(optional = true)
+  public transient DefaultInputPort<T> input2 = new DefaultInputPort<T>()
+  {
+    @Override
+    public void process(T tuple)
+    {
+      isLeft = false;
+      processTuple(tuple);
+    }
+  };
+  // Stores for each of the input port
+  private BackupStore[] store = (BackupStore[])Array.newInstance(BackupStore.class, 2);
+  private String includeFieldStr;
+  private String keyFieldStr;
+  private String timeFieldStr;
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    if (store[0] == null) {
+      throw new RuntimeException("Left Store is Empty");
+    }
+    if (store[1] == null) {
+      throw new RuntimeException("Right Store is Empty");
+    }
+    // Checks whether the strategy is outer join and set it to store
+    boolean isOuter = strategy.equals(JoinStrategy.LEFT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN);
+    store[0].isOuterJoin(isOuter);
+    isOuter = strategy.equals(JoinStrategy.RIGHT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN);
+    store[1].isOuterJoin(isOuter);
+    // Setup the stores
+    store[0].setup();
+    store[1].setup();
+
+    populateFields();
+  }
+
+  /**
+   * Create the event with the given tuple. If it successfully inserted it into the store
+   * then it does the join operation
+   *
+   * @param tuple Tuple to process
+   */
+  protected void processTuple(T tuple)
+  {
+    int idx = 0;
+    if (!isLeft) {
+      idx = 1;
+    }
+    TimeEvent t = createEvent(tuple);
+    if (store[idx].put(t)) {
+      join(t, isLeft);
+    }
+  }
+
+  private void populateFields()
+  {
+    populateIncludeFields();
+    populateKeyFields();
+    if (timeFieldStr != null) {
+      populateTimeFields();
+    }
+  }
+
+  /**
+   * Populate the fields from the includeFiledStr
+   */
+  private void populateIncludeFields()
+  {
+    includeFields = new String[2][];
+    String[] portFields = includeFieldStr.split(";");
+    for (int i = 0; i < portFields.length; i++) {
+      includeFields[i] = portFields[i].split(",");
+    }
+  }
+
+  /**
+   * Get the tuples from another store based on join constraint and key
+   *
+   * @param tuple  input
+   * @param isLeft whether the given tuple is from first port or not
+   */
+  private void join(TimeEvent tuple, Boolean isLeft)
+  {
+    // Get the valid tuples from the store based on key
+    // If the tuple is null means the join type is outer and return unmatched tuples from store.
+    Object value;
```

Comment:
if value is of type ArrayList<TimeEvent> joinedValues why declare as Object type?